ext/standard/password.c

```diff
@@ -356,30 +515,92 @@ PHP_FUNCTION(password_hash)
 		salt_len = required_salt_len;
 	}
 
-	salt[salt_len] = 0;
+	switch (algo) {
+		case PHP_PASSWORD_BCRYPT:
+			{
+				zend_string *result;
+				salt[salt_len] = 0;
 
-	hash = safe_emalloc(salt_len + hash_format_len, 1, 1);
-	sprintf(hash, "%s%s", hash_format, salt);
-	hash[hash_format_len + salt_len] = 0;
+				hash = safe_emalloc(salt_len + hash_format_len, 1, 1);
+				sprintf(hash, "%s%s", hash_format, salt);
+				hash[hash_format_len + salt_len] = 0;
 
-	efree(salt);
+				efree(salt);
 
-	/* This cast is safe, since both values are defined here in code and cannot overflow */
-	hash_len = (int) (hash_format_len + salt_len);
+				/* This cast is safe, since both values are defined here in code and cannot overflow */
+				hash_len = (int) (hash_format_len + salt_len);
 
-	if ((result = php_crypt(password, (int)password_len, hash, hash_len, 1)) == NULL) {
-		efree(hash);
-		RETURN_FALSE;
-	}
+				if ((result = php_crypt(password, (int)password_len, hash, hash_len, 1)) == NULL) {
+					efree(hash);
+					RETURN_FALSE;
+				}
 
-	efree(hash);
+				efree(hash);
 
-	if (ZSTR_LEN(result) < 13) {
-		zend_string_free(result);
-		RETURN_FALSE;
-	}
+				if (ZSTR_LEN(result) < 13) {
+					zend_string_free(result);
+					RETURN_FALSE;
+				}
 
-	RETURN_STR(result);
+				RETURN_STR(result);
+			}
+			break;
+#if HAVE_ARGON2LIB
+		case PHP_PASSWORD_ARGON2I:
+		case PHP_PASSWORD_ARGON2D:
+			{
+				char *out;
+				char *encoded;
+
+				size_t out_len = 32;
+				size_t encoded_len;
+				int status = 0;
+
+				encoded_len = argon2_encodedlen(
+					t_cost,
+					m_cost,
+					threads,
+					(uint32_t)salt_len,
+					out_len
+				);
+
+				encoded = emalloc(encoded_len + 1);
+				out = emalloc(out_len + 1);
+
+				status = argon2_hash(
+					t_cost,
+					m_cost,
+					threads,
+					password,
+					password_len,
+					salt,
+					salt_len,
+					out,
+					out_len,
+					encoded,
+					encoded_len,
+					type,
+					ARGON2_VERSION_NUMBER
+				);
+
+				zend_string *ret = zend_string_init(encoded, encoded_len, 0);
```

Comment:
Both out and encoded are converted to `zend_strings` now via `zend_string_alloc`, and `encoded` is properly freed if the error condition is hit now.